library/alloc/src/rc.rs

```diff
@@ -2776,8 +2776,16 @@ impl<T: ?Sized> Weak<T> {
     ///
     /// # Safety
     ///
-    /// The pointer must have originated from the [`into_raw`] and must still own its potential
-    /// weak reference, and `ptr` must point to a block of memory allocated by the global allocator.
+    /// The pointer must have originated from the [`into_raw`] or [`into_raw_and_alloc`] functions
+    /// and must still own its potential weak reference, and `ptr` must point to a block of
+    /// memory allocated by `alloc`.
+    ///
+    /// Note that `from_raw` expects values that actually originated from a call to one of these
+    /// functions, not what these functions can maybe return, or are documented to potentially return.
+    /// For example, [`into_raw`] can return dangling pointers, but this doesn't allow you to create
+    /// a dangling pointer yourself and pass it to `from_raw`. This rule holds even if that pointer
+    /// created via some other means has the same value than some output you have observed from
```

Comment:
If it has entirely the same value then you can use it. The subtle thing is that "same value" on pointers means "same address and same provenance".

What is that last sentence supposed to defend against?

Reply:
The first was to explain that even though `into_raw` might return *some* dangling pointer, including possibly the null pointer, `from_raw` does not allow you to create an arbitrary dangling pointer. The second was to explain that even if the address is the same (or whatever else goes into the `PartialEq`/`Hash` impls on pointers), it's still not allowed. I don't know how provenance is defined, and if the docs still say that [provenance's definition is vague](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#pointer-provenance), then I think explaining the impact here in clear words is helpful. I thought provenance was only a concept for *valid* pointers for example, didn't think that different invocations of `ptr::invalid_mut()` with the same address also yielded different provenances, instead of just having one and the same "INVALID" provenance (whereas proper allocations each create their own provenance).

Also, even if provenance allows people to use `ptr::invalid_mut(usize::MAX)`, which is equivalent to `Weak::into_inner(Weak::new())`, we still want to forbid that, because we reserve changing the implementation in the future, e.g. to use null pointers instead of `usize::MAX`. That's at least how I understand the current documentation.

Reply:
> The second was to explain that even if the address is the same (or whatever else goes into the PartialEq/Hash impls on pointers), it's still not allowed.

But the first already says that. It sounds like you are saying the same thing 3 different times using different words, and I still find that more confusing than helpful.

Reply:
Okay I'll remove the repetition, but I'll keep the part with `Weak::new` as @cuviper likes it.